service/storage_service.cc

```diff
@@ -5078,13 +5084,31 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
                     //co_await sleep_abortable(_db.local().get_config().read_request_timeout_in_ms() * std::chrono::milliseconds(1), _abort_source);
                     result.status = raft_topology_cmd_result::command_status::success;
                 break;
+                case raft_topology_cmd::command::fence: {
+                    // We can have several concurrent fence commands in case topology change
+                    // coordinator migrated to another node. The update_fence_version function
+                    // checks that the version doesn't decrease, we do the check and persist
+                    // the new version under the same lock to avoid raises.
+                    auto holder = co_await get_units(_raft_topology_cmd_handler_state._operation_mutex, 1);
+                    co_await update_fence_version(version);
```

Comment:
Since `version` is now defined above, before `get_units` which may yield, and the topology version may change during the yield`, `update_fence_version` may actually detect a mismatch, IIUC.

That's good, but it's no good that we use `on_internal_error` for a situation which is not a bug.
We should either update `update_fence_version` to not use `on_internal_error` or do another check here. If the check fails we should use a regular throw since it's a situation that may arise, not a bug.

Reply:
Fixed [here](https://github.com/gusev-p/scylla/tree/raft_fencing3_fix), will turn it to PR when this is merged.